pycparser/c_parser.py

```diff
@@ -1081,6 +1089,16 @@ def p_enumerator_list(self, p):
             p[1].enumerators.append(p[3])
             p[0] = p[1]
 
+    def p_atomic_specifier(self, p):
+        """ atomic_specifier  : _ATOMIC LPAREN type_name RPAREN
+        """
+        p[0] = c_ast.TypeDecl(
```

Comment:
I don't think you should be creating a new TypeDecl here. Instead you should be attaching a `['atomic']` qualifier on an existing declaration.

Reply:
Hmm, but do I have one? Or do you refer to `type_name`?

Reply:
With the latest change it correctly generates `int * _Atomic a;` from `_Atomic(int *) a;`, but their corresponding ASTs are normalised differently:

<details>
<summary>ASTs</summary>

```
FileAST(ext=[Decl(name='a',
                  quals=[
                        ],
                  storage=[
                          ],
                  funcspec=[
                           ],
                  type=PtrDecl(quals=['_Atomic'
                                     ],
                               type=TypeDecl(declname='a',
                                             quals=[
                                                   ],
                                             type=IdentifierType(names=['int'
                                                                       ]
                                                                 )
                                             )
                               ),
                  init=None,
                  bitsize=None
                  )
            ]
        )
```

```
FileAST(ext=[Decl(name='a',
                  quals=[
                        ],
                  storage=[
                          ],
                  funcspec=[
                           ],
                  type=TypeDecl(declname='a',
                                quals=[
                                      ],
                                type=Typename(name=None,
                                              quals=[
                                                    ],
                                              type=PtrDecl(quals=['_Atomic'
                                                                 ],
                                                           type=TypeDecl(declname=None,
                                                                         quals=[
                                                                               ],
                                                                         type=IdentifierType(names=['int'
                                                                                                   ]
                                                                                             )
                                                                         )
                                                           )
                                              )
                                ),
                  init=None,
                  bitsize=None
                  )
            ]
        )
```
</details>

I think the generated tree for `int * _Atomic a;` is correct, while `_Atomic(int *) a;` has some trash nodes that need to be removed. I thought this is what `_build_declarations` should fix, but actually it does not.

- Should these ASTs match in this form (i.e. when just `print`ing?
- Is there any centralised code that drops such empty nodes from the AST? If not, do I need to do it locally in some particular place?
- Should the comparison code in the test be fixed instead?

Reply:
An alternative would be to avoid AST comparison at all, checking only the resulting nodes for known configurations. I need your decision here as unlike with the `Atomic` node type there is no simple way to provide generation code.

Reply:
Sorry, I've been busy with other things.

Whenever I look at this, it's clear that it requires more thought than a cursory review. These parts of the parser are the trickiest. I'll need to find a significant chunk of time to look into this and upload all the required state back into my head.

Reply:
Sure. For the time being we have a workaround with the node. As long as you could spare some time eventually and provide some guidance, I believe we should be able to figure it out. Currently I simply do not have enough architectural context to provide a reasonable solution.